commit 21cfde2548d5c64753c27420366dc031f702eca2
Author: Prabha Murugesan <prabha.murugesan@morsemicro.com>
Date:   Wed Jun 11 18:34:41 2025 +1000

    [APP-4507][PR #40] Implement generate_connected_clients_events for netlink monitor hal

diff --git a/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp
index bd47dd123..04b238f93 100644
--- a/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp
@@ -511,24 +511,109 @@ bool mon_wlan_hal_nl80211::channel_scan_dump_results()
 bool mon_wlan_hal_nl80211::generate_connected_clients_events(
     bool &is_finished_all_clients, std::chrono::steady_clock::time_point max_iteration_timeout)
 {
-    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED";
-    is_finished_all_clients = true;
+#if defined(MORSE_MICRO)
+    std::string iface_name = get_iface_name();
+    std::string cmd;
+    parsed_obj_map_t reply;
+    int connected_time = 0;
+    std::string mac_addr;
+
+    if (iface_name.empty()) {
+        LOG(TRACE) << __func__ << ": Empty interface name";
+        return false;
+    }
+
+    LOG(TRACE) << __func__ << " entered for " << iface_name;
+
+    while (true) {
+        // if thread awake time is too long - return false (means there is more handling to be done on next wake-up)
+        if (std::chrono::steady_clock::now() > max_iteration_timeout) {
+            LOG(DEBUG)
+                << "Thread is awake too long - will continue on next wakeup, last handled sta:"
+                << m_prev_client_mac;
+            is_finished_all_clients = false;
+            return true;
+        }
+        reply.clear();
+
+        if (!m_queried_first) {
+            cmd = "STA-FIRST";
+        } else {
+            cmd = "STA-NEXT " + tlvf::mac_to_string(m_prev_client_mac);
+        }
+
+        if (!wpa_ctrl_send_msg(cmd, reply, iface_name)) {
+            LOG(ERROR) << cmd << " for " << iface_name << " failed";
+            if (m_queried_first) {
+                m_queried_first   = false;
+                m_prev_client_mac = beerocks::net::network_utils::ZERO_MAC;
+                return true;
+            }
+            // Failure on the first client for that VAP is certainly an error
+            return false;
+        }
+
+        m_queried_first = true;
+        if (reply.empty()) {
+            LOG(TRACE) << "cmd=" << cmd << " returned empty, reply=" << reply;
+            break;
+        }
+
+        std::string reply_str;
+        for (const auto &entry : reply) {
+            if (!reply_str.empty()) {
+                reply_str += "\n";
+            }
+            reply_str += entry.first + "=" + entry.second;
+        }
+
+        connected_time = beerocks::string_utils::stoi(reply["connected_time"]);
+        mac_addr       = reply["dot11RSNAStatsSTAAddress"];
+
+        LOG(DEBUG) << cmd << " reply for " << iface_name << " = \n"
+                   << reply_str << "\n\nconnected_time=" << connected_time << ", mac=" << mac_addr;
+
+        auto msg_buff = ALLOC_SMART_BUFFER(sizeof(sACTION_MONITOR_CLIENT_ASSOCIATED_NOTIFICATION));
+        auto msg = reinterpret_cast<sACTION_MONITOR_CLIENT_ASSOCIATED_NOTIFICATION *>(msg_buff.get());
+        if (!msg) {
+            LOG(TRACE) << __func__ << "Memory allocation failed!";
+            return false;
+        }
+        msg->mac    = tlvf::mac_from_string(reply["dot11RSNAStatsSTAAddress"]);
+        msg->vap_id = get_vap_id_with_bss(iface_name);
+        m_prev_client_mac = msg->mac;
+
+        if (m_handled_clients.find(m_prev_client_mac) != m_handled_clients.end()) {
+            LOG(DEBUG) << "already generated event for this client" << m_prev_client_mac;
+            continue;
+        }
+        // Add the message to the queue
+        m_handled_clients.insert(m_prev_client_mac);
+        event_queue_push(Event::STA_Connected, msg_buff);
+    }
 
-    // TODO: implement the API (PPM-1152)
-    // currently returning true even though not implemented in order not to break
-    // the flow if this HAL is used by any flow, since the API return value is checked by
-    // a common flow in the monitor.
+    m_prev_client_mac = beerocks::net::network_utils::ZERO_MAC;
+    m_handled_clients.clear();
+    m_queried_first                              = false;
+    connected_clients_events_generation_complete = true;
+    LOG(DEBUG) << "Finished to generate connected clients events for all clients";
+#else
+    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED!";
+#endif
+    is_finished_all_clients = true;
     return true;
 }
 
 bool mon_wlan_hal_nl80211::pre_generate_connected_clients_events()
 {
-    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED";
-
-    // TODO: implement the API (PPM-1152)
-    // currently returning true even though not implemented in order not to break
-    // the flow if this HAL is used by any flow, since the API return value is checked by
-    // a common flow in the monitor.
+#if defined(MORSE_MICRO)
+    m_prev_client_mac = beerocks::net::network_utils::ZERO_MAC;
+    m_handled_clients.clear();
+    m_queried_first                              = false;
+    connected_clients_events_generation_complete = false;
+#else
+    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED!";
+#endif
     return true;
 }
 
@@ -577,6 +662,12 @@ bool mon_wlan_hal_nl80211::process_nl80211_event(parsed_obj_map_t &parsed_obj)
         msg->vap_id = vap_id;
         msg->mac    = tlvf::mac_from_string(parsed_obj["_mac"]);
 
+#if defined(MORSE_MICRO)
+        // To prevent duplication of generation of connected event for clients,
+        // need to add associated clients to the "handled_clients" set
+        if (!connected_clients_events_generation_complete)
+            m_handled_clients.insert(msg->mac);
+#endif
         // Add the message to the queue
         event_queue_push(Event::STA_Connected, msg_buff);
 
diff --git a/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.h b/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.h
index fcd192d54..3b7b658ee 100644
--- a/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.h
+++ b/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.h
@@ -11,6 +11,9 @@
 
 #include "base_wlan_hal_nl80211.h"
 #include <bwl/mon_wlan_hal.h>
+#if defined(MORSE_MICRO)
+#include <bcl/network/network_utils.h>
+#endif
 
 namespace bwl {
 namespace nl80211 {
@@ -90,6 +93,12 @@ protected:
     // Private data-members:
 private:
     std::shared_ptr<char> m_temp_wav_value;
+#if defined(MORSE_MICRO)
+    std::unordered_set<sMacAddr> m_handled_clients;
+    sMacAddr m_prev_client_mac                        = beerocks::net::network_utils::ZERO_MAC;
+    bool m_queried_first                              = false;
+    bool connected_clients_events_generation_complete = false;
+#endif
 };
 
 } // namespace nl80211
